Support CUDA IPC in Python 3

colesbury opened the pull request:
CUDA IPC only works with Python 3 using the "spawn" start method. You can select the start method using the get_context method:

```python
 import torch.multiprocessing as mp
 ctx = mp.get_context('spawn')
 queue = ctx.Queue()
 event = ctx.Event()
```

To work better with multiprocessing contexts (which were introduced in Python 3), I've changed the way you switch the "sharing strategy". To use a non-default sharing strategy you get a new context:

```python
import torch.multiprocessing as mp
ctx = mp.get_context(sharing_strategy="file_system")
```

(This works in both Python 2 and Python 3)

apaszke reviewed (changes requested):
I also think it's a good idea to move the sharing part out of StorageMethods.cpp. It's large and complex enough that it deserves a separate file.

colesbury committed:
Support CUDA IPC in Python 3

CUDA IPC only works with Python 3 using the "spawn" start method. You
can select the start method using the get_context method:

 import torch.multiprocessing as mp
 ctx = mp.get_context('spawn')
 queue = ctx.Queue()
 event = ctx.Event()

apaszke reviewed (commented):
Mostly looks good, only some minor comments.

soumith reviewed (approved):
reviewed thoroughly
